src/libstore/gc.cc

```diff
@@ -17,8 +17,39 @@
 #include <unistd.h>
 #include <climits>
 
+
 namespace nix {
 
+struct IoNice {
+    enum {
+        IOPRIO_WHO_PROCESS = 1,
+        IOPRIO_WHO_PGRP,
+        IOPRIO_WHO_USER
+    };
+    enum {
+        IOPRIO_CLASS_NONE,
+        IOPRIO_CLASS_RT,
+        IOPRIO_CLASS_BE,
+        IOPRIO_CLASS_IDLE
+    };
+
+    enum {
+        IOPRIO_CLASS_SHIFT = 13
+    };
+    IoNice(int cl) {
```

Comment:
You may as well hardcode this parameter, because the error message expects it that way and we don't have a use case for another value than `IOPRIO_CLASS_IDLE`.

Reply:
yes

maybe whole nix (and nix-daemon) should be ionice'd.
creating derivations in /nix/store is io heavy and low priority too

Reply:
Nix can be used interactively or in a 'management' role though. Starving deployment on an I/O bottlenecked server could make a bad day much worse.

Reply:
I meant exactly the case of a " I/O bottlenecked server".
It is when Nix must idle to what server normally does in order not to put the server down

Reply:
Or it is when Nix must be usable in order to fix the I/O problem.
If your service must provide hard realtime guarantees, use a realtime operating system. If your service must provide soft realtime guarantees, being able to fix issues quickly is at least as important as the service itself. If you got yourself into a situation where it's I/O bottlenecked, that is a problem with your deployment, because the deployment should be designed to stay well within limits. Hence, to fix the problem with the server, you need to be able to deploy, and the sooner, the better. If you prioritize the production load, you'd have to shut down the production service, which is likely worse, because it is not obvious and not something you would typically dare to automate either.
So I don't think idle is a good default for nix-daemon. In the few cases where it does make sense, specifically when you know you won't need Nix to fix stuff, you can set IOSchedulingClass in NixOS' systemd options. However, if you don't need Nix to fix stuff, you probably don't need its daemon either.

Reply:
I observed LA > 30 on a cheap vps with spindle disk every time when Nix downloaded and unpacked packages during routing upgrade.
There was nothing realtime so I ignored that and forget.
And only now I realize that it could be avoided.